Fix postlayout subckt sims

mkosunen opened the pull request.

Commit:
num_cols fix

vlahtin committed:
Change failure to warning

alttiheikkinen committed:
Merge branch 'fix_postlayout_subckt_sims' of github.com:TheSystemDevelopmentKit/spice into fix_postlayout_subckt_sims

alttiheikkinen committed:
Add postlayout ckt parameter

vlahtin committed:
Merge remote-tracking branch 'origin/num_cols_fix' into fix_postlayout_subckt_sims

vlahtin committed:
Fix the number of columns addition

mkosunen committed:
Subcircuits is a list, therefore plural postlayout_subckts

sporrasm commented:
I just tested this. It does what it is supposed to do. However, setting self.postlayout = True while specifying self.postlayout_subckts doesn't export the subcircuit to the simulation path and hence nothing works. Reason for this is line 290 of  spice/spice_testbench.py (https://github.com/TheSystemDevelopmentKit/spice/blob/c7116c236bbf57dfa237bd67ad3e79df4bd5621f/spice/testbench.py#L290-L291)
 
 I need postlayout to be True to enable parastitic optimizations in Spectre. Does @mkosunen remember why you chose to enforce non-zero dspf file list length when self.postlayout is True? There may be other file formats that we wish to simulate as well, so I'm thinking that this shouldn't be the case.

sporrasm commented:
@mkosunen I commented the above cited lines from the code, set self.postlayout_subckts to some list and self.postlayout to True in sim and to False in the other. The simulation with postlayout flag set to True finished several hours before the one with postlayout flag set to False. The reason for this is parasitic reduction inside Spectre, which is enabled by setting postlayout to True (see plflag attribute in __init__.py).

Are the above lines really mandatory? According to https://github.com/TheSystemDevelopmentKit/spice/blob/c7116c236bbf57dfa237bd67ad3e79df4bd5621f/spice/__init__.py#L442-L445, self.postlayout is only to set the postlayout reduction flag. Is this not the case anymore or why non-zero dspf length enforced?

Commit:
Remove check that prevents exporting of subckt if postlayout is True

Commit:
Fix setting postlayout flag for spectre

Commit:
Add warning print that was removed in cc26a43ddc

The print is now a warning and notifies the user that both
dspf and postlayout_subckts were empty lists if postlayout
was True. The user is encouraged to take a look at the netlist
from spicesrc and check that is truly is postlayout.

sporrasm commented:
Re-added the warning print if postlayout is True and both dspfs lists are empty. As an added bonus, added a print requesting the user to check if self.spicesrc is truly a postlayout netlist.

Merging.
